File: docs/diagrama/gerar.py
"""Gera docs/diagrama.png a partir de docs/diagrama/diagrama.html.

    python docs/diagrama/gerar.py

Quem desenha de fato é o Electron (gerar.js), o mesmo que o app de bandeja já
usa. Este script só acha o binário e chama.

Por que não Chrome headless: nesta máquina o `--screenshot` sai SEM gerar nada
e SEM erro, e a porta de debug (`--remote-debugging-port`) nunca abre — nem com
perfil novo, nem com porta 0, e não há política do Chrome bloqueando. O
capturePage() do Electron não depende de porta nenhuma e funciona sempre.
"""
import subprocess
import sys
from pathlib import Path

AQUI = Path(__file__).resolve().parent
RAIZ = AQUI.parents[1]
SAIDA = AQUI.parent / "diagrama.png"
ELECTRON = RAIZ / "apps" / "desktop" / "node_modules" / "electron" / "dist" / "electron.exe"


def main():
    if not ELECTRON.is_file():
        print(f"não achei o Electron em {ELECTRON}")
        print("rode `npm install` em apps/desktop")
        return 2
    antes = SAIDA.stat().st_mtime if SAIDA.is_file() else 0
    r = subprocess.run([str(ELECTRON), str(AQUI)], capture_output=True, text=True,
                       timeout=180)
    saida = (r.stdout or "").strip()
    if saida:
        print(saida)
    if not SAIDA.is_file() or SAIDA.stat().st_mtime == antes:
        print("não gerou a imagem")
        print((r.stderr or "")[-600:])
        return 1
    return 0


sys.exit(main())

File: docs/diagrama/gerar.js
// Rasteriza docs/diagrama/diagrama.html em docs/diagrama.png.
//
// Roda pelo Electron que já vem com o app de bandeja:
//     apps/desktop/node_modules/.bin/electron docs/diagrama
// (o gerar.py chama isso pra você)
//
// Por que Electron e não Chrome headless: nesta máquina o `--screenshot` do
// Chrome sai sem gerar nada e sem erro, e a porta de debug (--remote-debugging-
// port) nunca abre — nem com perfil novo. capturePage() do Electron é
// direto e não depende de porta nenhuma.
const { app, BrowserWindow } = require('electron')
const fs = require('fs')
const path = require('path')

const LARGURA = 1680
const ALTURA = 812
const ESCALA = 2                       // 2x = legível em tela grande
const HTML = path.join(__dirname, 'diagrama.html')
const SAIDA = path.join(__dirname, '..', 'diagrama.png')

app.disableHardwareAcceleration()

app.whenReady().then(async () => {
  const win = new BrowserWindow({
    width: LARGURA, height: ALTURA, show: false,
    useContentSize: true,
    webPreferences: { offscreen: true, backgroundThrottling: false },
  })
  await win.loadFile(HTML)
  // deixa fonte, gradiente e emoji assentarem antes de capturar
  await new Promise(r => setTimeout(r, 1200))

  // capturePage() sozinho devolve o tamanho da JANELA, que o Windows limita
  // ao tamanho da tela (saía 1920 de largura em vez de 3360). Pelo debugger
  // dá pra fixar as medidas e a escala, independente do monitor.
  const dbg = win.webContents.debugger
  dbg.attach('1.3')
  await dbg.sendCommand('Emulation.setDeviceMetricsOverride', {
    width: LARGURA, height: ALTURA, deviceScaleFactor: ESCALA, mobile: false,
  })
  await new Promise(r => setTimeout(r, 400))
  const { data } = await dbg.sendCommand('Page.captureScreenshot', {
    format: 'png', captureBeyondViewport: true,
  })
  dbg.detach()

  const png = Buffer.from(data, 'base64')
  fs.writeFileSync(SAIDA, png)
  console.log(`${SAIDA}  (${Math.round(png.length / 1024)} KB, ` +
              `${LARGURA * ESCALA}x${ALTURA * ESCALA})`)
  win.destroy()
  app.quit()
}).catch(e => {
  console.error('falhou:', e.message)
  app.exit(1)
})
